### ISSUE TITLE
Problem Configuring Security Server in Development Environment with Ansible/LXD

### ISSUE THREAD

**Issue by jmhal** [nordic-institute/X-Road#910]
I am performing a clean install of the ansible lxd development environment inside a dual core virtual machine running Ubuntu 20.04 with 8GB of RAM and 30GB of disk space.

This is my lxd_hosts.txt:
```
#central servers (ubuntu lxd containers)
[cs_servers]
xroad-lxd-cs ansible_connection=lxd

#certification authority, time stamping authority and ocsp service server for testing purposes (ubuntu)
[ca_servers]
xroad-lxd-cs ansible_connection=lxd

#security servers (ubuntu lxd containers)
[ss_servers]
xroad-lxd-ss1 ansible_connection=lxd
xroad-lxd-ss2 ansible_connection=lxd

#container host
[lxd_servers]
localhost ansible_connection=local

#compilation host
[compile_servers]
localhost ansible_connection=local

### vars ###

[ss_servers:vars]
variant=vanilla
```

The _xroad_init.yml_ is left unchanged. I start the configuration process with this:

```
$ ansible-playbook -i hosts/lxd_hosts.txt xroad_init.yml
```

And everything seems to finish smoothly and I am left with three containers, as expected:

```
+---------------+---------+---------------------+-----------------------------------------------+-----------+-----------+
|     NAME      |  STATE  |        IPV4         |                     IPV6                      |   TYPE    | SNAPSHOTS |
+---------------+---------+---------------------+-----------------------------------------------+-----------+-----------+
| xroad-lxd-cs  | RUNNING | 10.83.51.42 (eth0)  | fd42:55ce:ef88:e724:216:3eff:fed1:f36c (eth0) | CONTAINER | 0         |
+---------------+---------+---------------------+-----------------------------------------------+-----------+-----------+
| xroad-lxd-ss1 | RUNNING | 10.83.51.217 (eth0) | fd42:55ce:ef88:e724:216:3eff:fe0b:7a34 (eth0) | CONTAINER | 0         |
+---------------+---------+---------------------+-----------------------------------------------+-----------+-----------+
| xroad-lxd-ss2 | RUNNING | 10.83.51.221 (eth0) | fd42:55ce:ef88:e724:216:3eff:fe9c:af7d (eth0) | CONTAINER | 0         |
+---------------+---------+---------------------+-----------------------------------------------+-----------+-----------+
```

I log into https://10.83.51.42:4000 and I am able to perform the steps outlined in [this](https://confluence.niis.org/pages/viewpage.action?pageId=6783483#HowtoConfigureCentralServer?-4.2AddsubsystemtoSecurityServer) with no problem at all. So at the end, I have a _Central Server_ set up and an anchor file. So I go to a _Security Server_ at https://10.83.51.217:4000 for uploading the anchor file. But before the file is fully loaded, the system logs out and I get the following message:

` Error: Request failed with status code 401
ID: ef95584078bb4d40 `

Inside the _Security Server_ container, the _/var/log/xroad/audit.log_ has the following content:
 
```
2021-03-15T18:08:40+00:00 xroad-lxd-ss1 correlation-id: [34edfc15e3316a55] INFO  [X-Road Proxy Admin REST API] 2021-03-15T18:08:40.999Z - {"event":"Log in user","user":"xrd","url":"/login","data":{}}
2021-03-15T18:09:12+00:00 xroad-lxd-ss1 correlation-id: [ef95584078bb4d40] INFO  [X-Road Proxy Admin REST API] 2021-03-15T18:09:12.718Z - {"event":"Auth credentials discovery failed","user":null,"reason":"An Authentication object was not found in the SecurityContext","warning":false,"url":"/api/v1/notifications/alerts","data":{}}
2021-03-15T18:09:12+00:00 xroad-lxd-ss1 correlation-id: [3a389abd53d119cd] INFO  [X-Road Proxy Admin REST API] 2021-03-15T18:09:12.720Z - {"event":"Auth credentials discovery failed","user":null,"reason":"An Authentication object was not found in the SecurityContext","warning":false,"url":"/api/v1/notifications/session-status","data":{}}
root@xroad-lxd-ss1:/var/log/xroad# cat audit.log
2021-03-15T18:08:40+00:00 xroad-lxd-ss1 correlation-id: [34edfc15e3316a55] INFO  [X-Road Proxy Admin REST API] 2021-03-15T18:08:40.999Z - {"event":"Log in user","user":"xrd","url":"/login","data":{}}
2021-03-15T18:09:12+00:00 xroad-lxd-ss1 correlation-id: [ef95584078bb4d40] INFO  [X-Road Proxy Admin REST API] 2021-03-15T18:09:12.718Z - {"event":"Auth credentials discovery failed","user":null,"reason":"An Authentication object was not found in the SecurityContext","warning":false,"url":"/api/v1/notifications/alerts","data":{}}
2021-03-15T18:09:12+00:00 xroad-lxd-ss1 correlation-id: [3a389abd53d119cd] INFO  [X-Road Proxy Admin REST API] 2021-03-15T18:09:12.720Z - {"event":"Auth credentials discovery failed","user":null,"reason":"An Authentication object was not found in the SecurityContext","warning":false,"url":"/api/v1/notifications/session-status","data":{}}
``` 

There are some strange messages about a missing Security Context object, but that's before I am able to upload the anchor file. The _configuration_client.log_ file has several repeated lines with:

```
2021-03-15T18:27:44.863Z [QuartzScheduler_Worker-2] INFO  e.r.x.c.c.g.ConfigurationClientMain$ConfigurationClientJobListener - job was executed result=DiagnosticsStatus(returnCode=125, prevUpdate=2021-03-15T18:27:44.862Z, nextUpdate=2021-03-15T18:28:44.862Z, description=null)
2021-03-15T18:28:44.862Z [QuartzScheduler_Worker-3] WARN  e.r.x.c.c.g.ConfigurationClient - Cannot download configuration, anchor file /etc/xroad/configuration-anchor.xml does not exist
2021-03-15T18:28:44.863Z [QuartzScheduler_Worker-3] INFO  org.quartz.core.JobRunShell - Job jobs.ConfigurationClientJob threw a JobExecutionException:
org.quartz.JobExecutionException: java.io.FileNotFoundException: /etc/xroad/configuration-anchor.xml
        at ee.ria.xroad.common.conf.globalconf.ConfigurationClientJob.execute(ConfigurationClientJob.java:65)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.io.FileNotFoundException: /etc/xroad/configuration-anchor.xml
        at ee.ria.xroad.common.conf.globalconf.ConfigurationClient.initConfigurationAnchor(ConfigurationClient.java:95)
        at ee.ria.xroad.common.conf.globalconf.ConfigurationClient.execute(ConfigurationClient.java:70)
        at ee.ria.xroad.common.conf.globalconf.ConfigurationClientJob.execute(ConfigurationClientJob.java:51)
        ... 2 common frames omitted
2021-03-15T18:28:44.863Z [QuartzScheduler_Worker-3] INFO  e.r.x.c.c.g.ConfigurationClientMain$ConfigurationClientJobListener - job was executed result=DiagnosticsStatus(returnCode=125, prevUpdate=2021-03-15T18:28:44.863Z, nextUpdate=2021-03-15T18:29:44.863Z, description=null)
```

Should I try to manually copy the anchor file created by the _Central Server_ to _/etc/xroad/configuration-anchor.xml_  in the _Security Server_ containers?

**Comment by petkivim** [comment]
Hi @jmhal 

Does the problem occur only on `.217` or is `.221` affected too?

You could try to upload the configuration anchor using the management REST API of the Security Server. First, you must [create an API key](
https://github.com/nordic-institute/X-Road/blob/develop/doc/Manuals/ug-ss_x-road_6_security_server_user_guide.md#1911-creating-new-api-keys), and then upload the configuration anchor using [this](https://github.com/nordic-institute/X-Road/blob/8b5f0dd88b127291023f5b8696f9becbd983006b/src/proxy-ui-api/src/main/resources/openapi-definition.yaml#L3958) endpoint.

Best regards,
Petteri

**Comment by jmhal** [comment]
Hi @petkivim ,

Both .217 and 221. were affected. I did upload the anchor as you recommended, but after tinkering around, I found out that using SSH proxy to access the endpoints running in the LXD hosts inside the virtual machine somehow makes the web interface of the Security Server unstable. I am doing the steps again, now with a desktop version of Ubuntu instead of the server edition, and the web interface is working flawlessly with the Firefox browser running inside the VM. 

Thank you for your reply.
